Support for (de-)serializing geometry_msgs/PoseStamped

fairlight1337 opened the pull request.

Commit:
send-goal-and-wait uses conditions and timers instead of loop sleep

Commit:
refactored wait-for-result and it is now used by send-goal-and-wait

Commit:
Errors with lost goals and running into execute timeout when the goal already finished seem to be fixed

Commit:
Fixed goal status returning to active after finishing and goals now are getting marked as lost after receiving 20 status msg without it

fairlight1337 committed:
Support for (de-)serializing geometry_msgs/PoseStamped

fairlight1337 commented:
Since multiple connected systems use poses instead of transforms, having a way to serialize them might be good.

These include
- `rviz` (Pose and PoseArray, both using PoseStamped internally)
- `semrec` (PoseStamped)
- `designator_integration` (PoseStamped)

In the long run, we can think about whether the second and third could be refactored to use transforms rather than poses, but at least for now that won't be an option.

fairlight1337 committed:
Export pose-stamped to msg and vice versa conversion functions

fairlight1337 commented on cl_tf2/src/cl-transforms-plugin/ros-msg-conversions.lisp:
```diff
@@ -51,6 +58,12 @@
     (cl-transforms:make-transform
      (msg->3d-vector translation) (msg->quaternion rotation))))
 
+(defun msg->pose (msg)
```
This function (and its counterpart, `pose->msg`) are right now only implemented in `cl-tf`, which we want to get rid of as far as I understand. That's why I re-implemented them here.

fairlight1337 committed:
More (de-)serialization methods

airballking commented on cl_tf2/src/cl-transforms-plugin/ros-msg-conversions.lisp:
```diff
@@ -84,6 +109,13 @@
+  (declare (type cl-transforms:pose pose))
+  (make-msg
+   "geometry_msgs/Pose"
+   :position (cl-transforms:origin pose)
```
I think you need to wrap `(cl-transforms:origin pose)` in a call to `(point->msg ...)`.

airballking commented on cl_tf2/src/cl-transforms-plugin/ros-msg-conversions.lisp:
```diff
@@ -84,6 +109,13 @@
+  (make-msg
+   "geometry_msgs/Pose"
+   :position (cl-transforms:origin pose)
+   :orientation (cl-transforms:orientation pose)))
```
I think you need to wrap `(cl-transforms:orientation pose)` in a call to `(quaternion->msg ...)`.

fairlight1337 replied:
Good point. Will fix that. Same for above.

airballking commented:
I'm fine with the proposition. See my inline code review above.

fairlight1337 committed:
Added an additional copy constructor that directly exposes the slots sub-slots

This reflects the `old' behavior in which subslots of e.g. the header and pose of a pose-stamped could be set directly. Usage:

```lisp
> (copy-ext-pose-stamped pose-stamped :stamp 10.0
                                      :origin (cl-transforms:make-3d-vector 1 2 3))
```

Minor quirk: Due to type assertions in the function, the header's `stamp' field only accepts floats, no integers.

fairlight1337 committed:
Fixed a function name typo, and fixed serialization `pose->msg' as suggested by @airballking

fairlight1337 commented:
Okay, addressed all the issues you commented on. Should work now. I'll refactor the rest of CRAM based on the new convenience copy constructor, as that makes things a hell lot easier.

Also, proposals; you're right.

fairlight1337 committed:
Automatic export for stamped functions, bug fixes, doc

Make sure that copy classes are finalized before reading their slots; also, automatically export copy constructors and `make-*' functions; plus documentation (a little)

fairlight1337 commented on cl_tf2/src/cl-transforms-plugin/package.lisp:
```diff
@@ -36,4 +36,4 @@
    ;; point-stamped
    point-stamped make-point-stamped copy-point-stamped point
    ;; pose-stamped
-   pose-stamped make-pose-stamped copy-pose-stamped pose))
```
This gets exported automatically inside of `def-stamped` now.

airballking replied:
As discussed earlier today, I'd like to have this explicitedly exported in the `package.lisp`. Like this, there is one place where all the exported API can be found by people.

fairlight1337 replied:
Fixed in https://github.com/fairlight1337/roslisp_common/commit/f81277eda159271cb3fe4c93f7ced4632a3d3d08.

fairlight1337 commented on cl_tf2/src/cl-tf2/stamped-data.lisp:
```diff
@@ -71,6 +71,7 @@
    - define a class 'number-stamped' with slots 'num-value' and 'header'
    - define a constructor-function 'make-number-stamped'
    - define a function 'copy-number-stamped'
+   - define a function 'copy-ext-number-stamped'
```
This was the best name I could think of. Maybe we can change `copy-...` to be a method?
Then, we can have multiple `copy-...` functions with different signatures. Of course, then the syntax is not being displayed in the log window anymore.. or is it?

Anyway, having it as an overloaded method somehow feels better than this `-ext-`-stuff. I'll try that out when I have time.

fairlight1337 replied:
Tried it today. Apparently,

``` lisp
(defgeneric method-name (param-1 param-2 &key &allow-other-keys)
  (:method (param-1 param-2 &key key-1 key-2)
    ...)
  (:method (param-1 param-2 &key key-3)
    ...)
```

doesn't work, as it will default to the first method defined. Thus, these methods are not differentiated by the keys specified when calling `method-name`.

I'll therefore for now stick to `copy-ext-...`.

fairlight1337 committed:
Make sure the correct types are fed into stamped-related functions

fairlight1337 committed:
Made code more elegant (removed duplicate loops) and added type-checks

fairlight1337 committed:
Typo: `quaterion` -> `quaternion`

fairlight1337 committed:
Creation of stamped converters for `def-stamped`

This way, functions like `pose->pose-stamped` get created and exporrted automatically.

fairlight1337 committed:
Stamped converter functions don't use keywords, but actual (non-optional) params

Fixed this issue

fairlight1337 committed:
Order of parameters for stamped converters, and type checks

airballking commented on cl_tf2/src/cl-tf2/stamped-data.lisp:
```diff
@@ -70,17 +70,29 @@
  This call will:
    - define a class 'number-stamped' with slots 'num-value' and 'header'
    - define a constructor-function 'make-number-stamped'
+   - define a stamped converter 'number->number-stamped'
```
As discussed earlier today, this should already be covered by the original `copy-number-stamped`.

fairlight1337 replied:
Fixed in https://github.com/fairlight1337/roslisp_common/commit/c768f72044f0d95a555b8d1c200f2bab2ca7c56d.

Commit:
fixed: one closing bracket too much

fairlight1337 committed:
Removed stamped converter function from `def-stamped` macro again, as its functionality is already 100% given by the `make-...` function.

fairlight1337 committed:
Also, updated docs

fairlight1337 committed:
Export manually again

fairlight1337 committed:
Forgot to add `pose`

fairlight1337 committed:
Type safety in method definition

fairlight1337 committed:
Conversion of pose-stamped to point-stamped message

fairlight1337 committed:
Introduce ensure-fully-qualified-name from cl-tf

fairlight1337 committed:
Export of msg transform stamped functions

fairlight1337 committed:
More type safety for stamped-generated functions

fairlight1337 committed:
Merge branch 'master' of https://github.com/jannikb/roslisp_common into tf2-refactoring

fairlight1337 committed:
NIL is okay for parameters when their *-p component is not set
